SteamKit2/SteamKit2/Util/HardwareUtils.cs

```diff
@@ -0,0 +1,398 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Management;
+using System.Net.NetworkInformation;
+using System.Text;
+using Microsoft.Win32;
+using SteamKit2.Util.MacHelpers;
+
+using static SteamKit2.Util.MacHelpers.LibC;
+using static SteamKit2.Util.MacHelpers.CoreFoundation;
+using static SteamKit2.Util.MacHelpers.DiskArbitration;
+using static SteamKit2.Util.MacHelpers.IOKit;
+using System.Threading.Tasks;
+
+namespace SteamKit2
+{
+    abstract class MachineInfoProvider
+    {
+        public static MachineInfoProvider GetProvider()
+        {
+            switch ( Environment.OSVersion.Platform )
+            {
+                case PlatformID.Win32NT:
+                case PlatformID.Win32Windows:
+                    return new WindowsInfoProvider();
+
+                case PlatformID.Unix:
+                    if ( Utils.IsRunningOnDarwin() )
+                    {
+                        return new OSXInfoProvider();
+                    }
+                    else
+                    {
+                        return new LinuxInfoProvider();
+                    }
+            }
+
+            return new DefaultInfoProvider();
+        }
+
+        public abstract byte[] GetMachineGuid();
+        public abstract byte[] GetMacAddress();
+        public abstract byte[] GetDiskId();
+    }
+
+    class DefaultInfoProvider : MachineInfoProvider
+    {
+        public override byte[] GetMachineGuid()
+        {
+            return Encoding.UTF8.GetBytes( Environment.MachineName + "-SteamKit" );
+        }
+
+        public override byte[] GetMacAddress()
+        {
+            // mono seems to have a pretty solid implementation of NetworkInterface for our platforms
+            // if it turns out to be buggy we can always roll our own and poke into /sys/class/net on nix
+
+            var firstEth = NetworkInterface.GetAllNetworkInterfaces()
+                .Where( i => i.NetworkInterfaceType == NetworkInterfaceType.Ethernet )
```

Comment:
What about other network interface types?

On OS X Mono 4.0.1, almost everything is showing as `Ethernet`.
On Windows .NET Framework, I'm seeing `Ethernet` for wired connections and `Wireless80211` for WiFi.

This seems like it could fall back to `"SteamKit-MacAddress"` a bit too frequently. What adapters does Steam search for?